tests/unit/test_tls_lib.py

```diff
@@ -0,0 +1,311 @@
+        # assume relation exists with a current certificate
+        self.relate_to_tls_certificates_operator()
+        self.harness.charm.app_peer_data["chain"] = "app-chain-old"
+        self.harness.charm.app_peer_data["cert"] = "app-cert-old"
+        self.harness.charm.app_peer_data["csr"] = "app-crs-old"
+        self.harness.charm.app_peer_data["ca"] = "app-ca-old"
+        self.harness.charm.unit_peer_data["cert"] = "unit-cert"
+
+        self.charm.tls.certs.on.certificate_available.emit(
+            certificate_signing_request="app-crs",
+            chain=["app-chain"],
+            certificate="app-cert",
+            ca="app-ca",
+        )
+
+        self.assertEqual(self.harness.charm.app_peer_data["chain"], "app-chain-old")
+        self.assertEqual(self.harness.charm.app_peer_data["cert"], "app-cert-old")
+        self.assertEqual(self.harness.charm.app_peer_data["ca"], "app-ca-old")
+
+        restart_mongod_service.assert_not_called()
+
+    # Helper functions
+    def relate_to_tls_certificates_operator(self) -> int:
+        """Relates the charm to the TLS certificates operator."""
+        rel_id = self.harness.add_relation(RELATION_NAME, "tls-certificates-operator")
+        self.harness.add_relation_unit(rel_id, "tls-certificates-operator/0")
+        return rel_id
+
+    def verify_external_rsa_csr(
+        self, specific_rsa=False, expected_rsa=None, specific_csr=False, expected_csr=None
+    ):
+        """Verifies values of external rsa and csr.
+
+        Checks if rsa/csr were randomly generated or if they are a provided value.
+        """
+        unit_rsa_key = self.harness.charm.unit_peer_data.get("key", None)
+        unit_csr = self.harness.charm.unit_peer_data.get("csr", None)
+        if specific_rsa:
+            self.assertEqual(unit_rsa_key, expected_rsa)
+        else:
+            self.assertEqual(unit_rsa_key.split("\n")[0], "-----BEGIN RSA PRIVATE KEY-----")
+
+        if specific_csr:
+            self.assertEqual(unit_csr, expected_csr)
+        else:
+            self.assertEqual(unit_csr.split("\n")[0], "-----BEGIN CERTIFICATE REQUEST-----")
+
+    def verify_internal_rsa_csr(
+        self, specific_rsa=False, expected_rsa=None, specific_csr=False, expected_csr=None
+    ):
+        """Verifies values of internal rsa and csr.
+
+        Checks if rsa/csr were randomly generated or if they are a provided value.
+        """
+        app_rsa_key = self.harness.charm.app_peer_data.get("key", None)
+        app_csr = self.harness.charm.app_peer_data.get("csr", None)
+        if specific_rsa:
+            self.assertEqual(app_rsa_key, expected_rsa)
+        else:
+            self.assertEqual(app_rsa_key.split("\n")[0], "-----BEGIN RSA PRIVATE KEY-----")
+
+        if specific_csr:
+            self.assertEqual(app_csr, expected_csr)
+        else:
+            self.assertEqual(app_csr.split("\n")[0], "-----BEGIN CERTIFICATE REQUEST-----")
```

Comment:
Huge nit, these could be the one method right? Something like:

```python
def verify_rsa_csr(
    self,
    specific_rsa=False,
    expected_rsa=None,
    specific_csr=False,
    expected_csr=None,
    object: Union[Unit, Application]
):
    rsa_key = self.harness.charm.model.get_relation(PEER).data[object].get("key", None)
    ...
```

Reply:
it could be, but I like it better like this. If this is a make-or-break nit I am happy to refactor otherwise I will leave as is : )

Reply:
Nah, it's a big nit.